src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,889 @@
+/*
+ * Copyright (c) 2024, SAP SE. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "jfr/recorder/service/jfrOptionSet.hpp"
+#include "jfr/periodic/sampling/jfrCPUTimeThreadSampler.hpp"
+#include "jfr/recorder/service/jfrEvent.hpp"
+#include "jfr/recorder/stacktrace/jfrAsyncStackTrace.hpp"
+#include "jfr/utilities/jfrTypes.hpp"
+#include "jfr/recorder/jfrRecorder.hpp"
+#include "jfr/periodic/sampling/jfrCallTrace.hpp"
+#include "jfr/recorder/storage/jfrBuffer.hpp"
+#include "jfr/utilities/jfrTime.hpp"
+#include "jfrfiles/jfrEventClasses.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/threadCrashProtection.hpp"
+#include "runtime/osThread.hpp"
+
+#if defined(LINUX)
+#include "signals_posix.hpp"
+
+const int64_t AUTOADAPT_INTERVAL_MS = 100;
+
+enum JfrSampleType {
+  // no sample, because thread not in walkable state
+  NO_SAMPLE = 0,
+  // sample from thread while in Java
+  JAVA_SAMPLE = 1,
+  // sample from thread while in native
+  NATIVE_SAMPLE = 2
+};
+
+static bool thread_state_in_java(JavaThread* thread) {
+  assert(thread != nullptr, "invariant");
+  switch(thread->thread_state()) {
```

Comment:
Why a full switch when only two states are of interest?

Reply:
Similarity to the other sampler.

Reply:
And it's far easier to read.

Reply:
I certainly disagree on the readability aspect. The switch means I have to look through each section to figure out what it does for each state. In contrast:
```
state = ...
return state == _thread_in_Java || state == _thread_in_Java_trans;
``
is IMO blindingly obvious.

Reply:
But this doesn't check whether the state could be something new an unsupported.

Reply:
I changed it to your recommendation.

Reply:
Consistency with existing code may take precedence here.